bundles/org.openhab.binding.opensprinkler/src/main/java/org/openhab/binding/opensprinkler/internal/api/OpenSprinklerApi.java

```diff
@@ -89,40 +100,153 @@ public abstract void openStation(int station, BigDecimal duration)
+
+    /**
+     * Returns true if the internal programs are allowed to auto start.
+     *
+     * @return true if enabled
+     */
+    public abstract boolean getIsEnabled();
+
+    public abstract void enablePrograms(Command command) throws UnauthorizedApiException, CommunicationApiException;
 
     /**
      * Returns the water level in %.
      *
      * @return waterLevel in %
-     * @throws CommunicationApiException
-     * @throws
      */
-    public abstract int waterLevel() throws CommunicationApiException;
+    public abstract int waterLevel();
 
     /**
      * Returns the number of total stations that are controllable from the OpenSprinkler
      * device.
      *
      * @return Number of stations as an int.
-     * @throws Exception
      */
-    public abstract int getNumberOfStations() throws Exception;
+    public abstract int getNumberOfStations();
 
     /**
      * Returns the firmware version number.
      *
      * @return The firmware version of the OpenSprinkler device as an int.
      * @throws Exception
      */
-    public abstract int getFirmwareVersion() throws CommunicationApiException;
+    public abstract int getFirmwareVersion() throws CommunicationApiException, UnauthorizedApiException;
+
+    /**
+     * Sends all the GET requests and stores/cache the responses for use by the API to prevent the need for multiple
+     * requests.
+     *
+     * @throws CommunicationApiException
+     * @throws UnauthorizedApiException
+     */
+    public abstract void refresh() throws CommunicationApiException, UnauthorizedApiException;
+
+    /**
+     * Ask the OpenSprinkler for the program names and store these for future use in a List.
+     *
+     * @throws CommunicationApiException
+     * @throws UnauthorizedApiException
+     */
+    public abstract void getProgramData() throws CommunicationApiException, UnauthorizedApiException;
+
+    /**
+     * Returns a list of all internal programs as a list of StateOptions.
+     *
+     * @return List<StateOption>
+     */
+    public abstract List<StateOption> getPrograms();
+
+    /**
+     * Return a list of all the stations the device has as List of StateOptions
+     *
+     * @return List<StateOption>
+     */
+    public abstract List<StateOption> getStations();
+
+    /**
+     * Runs a Program that is setup and stored inside the OpenSprinkler
+     *
+     * @param Program index number that you wish to run.
+     *
+     * @throws CommunicationApiException
+     * @throws UnauthorizedApiException
+     */
+    public abstract void runProgram(Command command) throws CommunicationApiException, UnauthorizedApiException;
+
+    /**
+     * Fetch the station names and place them in a list of List<StateOption>.
+     * Use getStations() to retrieve this list.
+     *
+     * @throws CommunicationApiException
+     * @throws UnauthorizedApiException
+     */
+    public abstract JnResponse getStationNames() throws CommunicationApiException, UnauthorizedApiException;
```

Comment:
JnResponse shouldn't be declared in a subclass if you are going to use it in the parent class. Please define JnResponse in this class instead.

Reply:
DONE